src/passkey_child/passkey_child_credentials.c

```diff
@@ -0,0 +1,322 @@
+*/
+
+#include <termios.h>
+#include <stdio.h>
+
+#include "util/crypto/sss_crypto.h"
+#include "util/debug.h"
+#include "util/util.h"
+
+#include "passkey_child.h"
+
+errno_t
+prepare_credentials(const struct passkey_data *data, fido_cred_t *cred) {
+    unsigned char userid[32];
+    unsigned char cdh[32];
+    errno_t ret = EOK;
+
+    ret = fido_cred_set_type(cred, data->type);
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_cred_set_type failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
+
+    ret = sss_generate_csprng_buffer(cdh, sizeof(cdh));
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "sss_generate_csprng_buffer failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
+
+    ret = fido_cred_set_clientdata_hash(cred, cdh, sizeof(cdh));
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "fido_cred_set_clientdata_hash failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
+
+    DEBUG(SSSDBG_TRACE_FUNC, "Setting Relying Party ID and name to %s.\n",
+          data->domain);
+
+    ret = fido_cred_set_rp(cred, data->domain, data->domain);
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_cred_set_rp failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
+
+    ret = sss_generate_csprng_buffer(userid, sizeof(userid));
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "sss_generate_csprng_buffer failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
+
+    DEBUG(SSSDBG_TRACE_FUNC, "Setting user: %s.\n", data->shortname);
+
+    ret = fido_cred_set_user(cred, userid, sizeof(userid), data->shortname,
+                             NULL, NULL);
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_cred_set_user failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
+
+    // Set to FIDO_OPT_OMIT instead of FIDO_OPT_FALSE for compatibility reasons
+    ret = fido_cred_set_rk(cred, FIDO_OPT_OMIT);
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_cred_set_rk failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
+
+done:
+    return ret;
+}
+
+ssize_t
+read_pin(char **pin)
+{
+    char *line_ptr = NULL;
+    struct termios old, new;
+    size_t line_len = 0;
+    ssize_t bytes_read;
+    ssize_t ret;
+
+    ret = tcgetattr(STDIN_FILENO, &old);
+    if (ret != 0) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Unable get the parameters associated with stdin.\n");
+        goto done;
+    }
+    new = old;
+    new.c_lflag &= ~ECHO;
+    ret = tcsetattr(STDIN_FILENO, TCSAFLUSH, &new);
+    if (ret != 0) {
+        DEBUG(SSSDBG_OP_FAILURE, "Unable to turn echoing off.\n");
+        goto done;
+    }
+
+    printf("Enter PIN: ");
```

Comment:
Can this be printed to stderr instead of stdout? It makes it easier when another process forks `passkey register`, stdout would only show the resulting passkey:id,key and can be pipe'd, stderr would show all the messages including "Enter PIN" or "Please touch the device".

Reply:
Done. Thanks for the feedback.

Reply:
Hi,

my concern with using stderr is that the calling process has to assume that there are errors when something is returned on stderr. So the handle this case where a calling process will send the PIN and prompt the user to touch the device a `--quite` option for `passkey_child` might be better.

Btw, it looks like you currently cannot send the PIN via stdin, at least `echo PIN | passkey_child ...` does not work for me. But this can be fixed later imo.

bye,
Sumit

Reply:
I noted the feedback and I'll work on it later. I'll open another PR to solve this problem.